src/WebJobs.Script.WebHost/Management/LegionInstanceManager.cs

```diff
@@ -0,0 +1,146 @@
+                return false;
+            }
+
+            if (context.IsWarmupRequest)
+            {
+                // No need to download since file already exists
+                return true;
+            }
+
+            if (_assignmentContext == null)
+            {
+                lock (_assignmentLock)
+                {
+                    if (_assignmentContext != null)
+                    {
+                        return _assignmentContext.Equals(context);
+                    }
+                    _assignmentContext = context;
+                }
+
+                _logger.LogInformation($"Starting Assignment. Cloud Name: {_environment.GetCloudName()}");
+
+                // set a flag which will cause any incoming http requests to buffer
+                // until specialization is complete
+                // the host is guaranteed not to receive any requests until AFTER assign
+                // has been initiated, so setting this flag here is sufficient to ensure
+                // that any subsequent incoming requests while the assign is in progress
+                // will be delayed until complete
+                _webHostEnvironment.DelayRequests();
+
+                // start the specialization process in the background
+                Task.Run(async () => await Assign(context));
+
+                return true;
+            }
+
+            // No lock needed here since _assignmentContext is not null when we are here
+            return _assignmentContext.Equals(context);
+        }
+
+        public Task<string> ValidateContext(HostAssignmentContext assignmentContext)
+        {
+            // Don't need to validate RunFromPackageContext in Legion
+            return null;
+        }
+
+        private async Task Assign(HostAssignmentContext assignmentContext)
+        {
+            try
+            {
+                // first make all environment and file system changes required for
+                // the host to be specialized
+                _logger.LogInformation("Applying {count} app setting(s)", assignmentContext.Environment.Count);
+                assignmentContext.ApplyAppSettings(_environment, _logger);
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, "Assign failed");
+                await _meshServiceClient.NotifyHealthEvent(ContainerHealthEventType.Fatal, GetType(), "Assign failed");
+                throw;
+            }
+            finally
+            {
+                // all assignment settings/files have been applied so we can flip
+                // the switch now on specialization
+                // even if there are failures applying context above, we want to
+                // leave placeholder mode
+                _logger.LogInformation("Triggering specialization");
+                _webHostEnvironment.FlagAsSpecializedAndReady();
+                _webHostEnvironment.ResumeRequests();
```

Comment:
If assign fails, the `_webHostEnvironment.DelayRequests()` will never be undone - is that intentional?

Reply:
If assign fails, we mark the container unhealthy and it will marked for recycling and be disposed. So, I don't think it will matter.